File: AdventOfCode/2024/17.2.py
import re


input = """Register A: 51064159
Register B: 0
Register C: 0

Program: 2,4,1,5,7,5,1,6,0,3,4,6,5,5,3,0"""
sample_input = """Register A: 2024
Register B: 0
Register C: 0

Program: 0,3,5,4,3,0"""

regex = r'Register A: (\d+)\nRegister B: (\d+)\nRegister C: (\d+)\n\nProgram: (.*)'
match = re.match(regex, input)
assert match is not None

register_a = int(match.group(1))
register_b = int(match.group(2))
register_c = int(match.group(3))
program = list(map(int, match.group(4).split(',')))


def run(register_a, register_b, register_c, program, log=False):
    instruct_ptr = 0
    out = []

    while instruct_ptr < len(program) - 1:
        instruct, org_value = program[instruct_ptr], program[instruct_ptr + 1]
        instruct_ptr += 2

        if log:
            print(f'instruct = {instruct}, org_value = {org_value}', end=' ')
        if 0 <= org_value <= 3:
            if log:
                print(f'value = {org_value}')
            value = org_value
        elif org_value == 4:
            if log:
                print('value = A')
            value = register_a
        elif org_value == 5:
            if log:
                print('value = B')
            value = register_b
        elif org_value == 6:
            if log:
                print('value = C')
            value = register_c
        else:
            return [-1]

        if instruct == 0:
            if log:
                print('A = A >> value')
            register_a //= 2**value
        elif instruct == 1:
            if log:
                print('B = B ^ org_value')
            register_b = register_b ^ org_value
        elif instruct == 2:
            if log:
                print('B = value % 8')
            register_b = value % 8
        elif instruct == 3:
            if log:
                print('JMP')
            if register_a != 0:
                instruct_ptr = org_value
        elif instruct == 4:
            if log:
                print('B = B ^ C')
            register_b = register_b ^ register_c
        elif instruct == 5:
            if log:
                print('OUT value % 8')
            out.append(value % 8)
        elif instruct == 6:
            if log:
                print('B = A >> value')
            register_b = register_a // 2**value
        elif instruct == 7:
            if log:
                print('C = A >> value')
            register_c = register_a // 2**value
    return out


run(register_a, register_b, register_c, program, log=True)


solves = set()


def dfs(cur: int, num_bits: int, program_index: int):
    for i in range(2 << 3):
        a = (i << num_bits) + cur

        run_res = run(a, register_b, register_c, program)

        if run_res == program:
            solves.add(a)

        equal_count = 0
        for r, p in zip(run_res, program):
            if r == p:
                equal_count += 1
            else:
                break

        if equal_count >= program_index:
            dfs(a, num_bits + 3, program_index + 1)


for i in range(2 << 12):
    if run(i, register_b, register_c, program)[0] == program[0]:
        dfs(i, 12, 1)

print('Solution:', min(solves))

File: AdventOfCode/2024/17.2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

// Define constants
#define PROGRAM_LENGTH 16
#define MAX_OUTPUT_SIZE PROGRAM_LENGTH
#define MAX_I 10000000000ULL  // 10,000,000,000

// Function to perform the "run" logic
void run(uint64_t register_a, uint64_t register_b, uint64_t register_c, int program[], int program_length, int *out, int *out_length) {
    uint64_t instr_ptr = 0;
    *out_length = 0;

    while (instr_ptr < (uint64_t)(program_length - 1)) {
        int instruct = program[instr_ptr];
        int org_value = program[instr_ptr + 1];
        instr_ptr += 2;

        int value;
        if (org_value >= 0 && org_value <= 3) {
            value = org_value;
        } else if (org_value == 4) {
            value = (int)register_a;
        } else if (org_value == 5) {
            value = (int)register_b;
        } else if (org_value == 6) {
            value = (int)register_c;
        } else {
            fprintf(stderr, "Invalid org_value: %d\n", org_value);
            exit(EXIT_FAILURE);
        }

        switch (instruct) {
            case 0:
                if (value >= 64) {
                    fprintf(stderr, "Shift value too large for instruction 0: %d\n", value);
                    exit(EXIT_FAILURE);
                }
                register_a /= (1ULL << value);
                break;
            case 1:
                register_b ^= (uint64_t)org_value;
                break;
            case 2:
                register_b = value % 8;
                break;
            case 3:
                if (register_a != 0) {
                    if (org_value >= program_length) {
                        fprintf(stderr, "Jump to invalid instruction pointer: %d\n", org_value);
                        exit(EXIT_FAILURE);
                    }
                    instr_ptr = org_value;
                }
                break;
            case 4:
                register_b ^= register_c;
                break;
            case 5:
                if (*out_length >= PROGRAM_LENGTH) {
                    return;
                }
                out[*out_length] = value % 8;
                (*out_length)++;
                break;
            case 6:
                if (value >= 64) {
                    fprintf(stderr, "Shift value too large for instruction 6: %d\n", value);
                    exit(EXIT_FAILURE);
                }
                register_b = register_a / (1ULL << value);
                break;
            case 7:
                if (value >= 64) {
                    fprintf(stderr, "Shift value too large for instruction 7: %d\n", value);
                    exit(EXIT_FAILURE);
                }
                register_c = register_a / (1ULL << value);
                break;
            default:
                fprintf(stderr, "Invalid instruction: %d\n", instruct);
                exit(EXIT_FAILURE);
        }
    }
}

int main(const int argc, const char *argv[]) {
    // Initialize registers based on the provided input
    uint64_t initial_register_a = 51064159;
    uint64_t initial_register_b = 0;
    uint64_t initial_register_c = 0;

    // Define the program
    int program[PROGRAM_LENGTH] = {2,4,1,5,7,5,1,6,0,3,4,6,5,5,3,0};

    // Prepare output buffer
    int out[MAX_OUTPUT_SIZE];
    int out_length = 0;

    // read in i min and i max from sys args
    uint64_t i_min = 0;
    uint64_t i_max = 0;

    if (argc == 3) {
        i_min = strtoull(argv[1], NULL, 10);
        i_max = strtoull(argv[2], NULL, 10);
    } else {
        fprintf(stderr, "Usage: %s <i_min> <i_max>\n", argv[0]);
        return 1;
    }

    // Iterate over possible values of 'i'
    for (uint64_t i = i_min; i < i_max; i++) {
        // Optional: Print progress every 1 million iterations
        if (i % 1000000 == 0 && i != 0) {
            // printf("Checked up to i = %llu\n", i);
        }
        // Run the program with current 'i' as Register A
        run(i, initial_register_b, initial_register_c, program, PROGRAM_LENGTH, out, &out_length);

        // Compare the output with the program
        // First, check if lengths match
        if (out_length != PROGRAM_LENGTH) {
            out_length = 0;  // Reset for next iteration
            continue;
        }

        // Then, check each element
        int match = 1;
        for (int j = 0; j < PROGRAM_LENGTH; j++) {
            if (out[j] != program[j]) {
                match = 0;
                break;
            }
        }

        if (match) {
            printf("Found matching i: %llu\n", i);
            return 0;
        }

        // Reset output length for next iteration
        out_length = 0;
    }

    printf("No matching 'i' found up to %llu\n", MAX_I);
    return 0;
}
